### PR TITLE
Type-check tokenizer/index.js

### PR DESCRIPTION
**Pull request by andy-ms**
| Q                 | A
| ----------------- | ---
| Bug fix?          | no
| Breaking change?  | no
| New feature?      | no
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | tests pass
| Fixed tickets     | comma-separated list of tickets fixed by the PR, if any
| License           | MIT

This adds type annotations to every method in `tokenizer/index.js` and enables type checking in the file.
This requires a few forward declarations. If desired, these could be eliminated later by defining the methods in a superclass to `Tokenizer` or in `Tokenizer` itself.

### CODE REVIEW

**Comment by hzoo** [comment]
Hey @andy-ms! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node v4 - linux](https://travis-ci.org/babel/babylon/jobs/220615354)
* [Node v6 - linux](https://travis-ci.org/babel/babylon/jobs/220615357)
* [Node v7 - linux](https://travis-ci.org/babel/babylon/jobs/220615360)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/220615362)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/220615365)

**Comment by hzoo** [comment]
Hey @andy-ms! It looks like one or more of your builds have failed. I've copied the relevant info below to save you some time.

* [Node v4 - linux](https://travis-ci.org/babel/babylon/jobs/220619094)
* [Node v6 - linux](https://travis-ci.org/babel/babylon/jobs/220619095)
* [Node v7 - linux](https://travis-ci.org/babel/babylon/jobs/220619096)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/220619097)
* [Node vlts/* - linux](https://travis-ci.org/babel/babylon/jobs/220619100)

File: src/tokenizer/index.js [hunk 1]
```diff
@@ -40,14 +44,28 @@ function codePointToString(code) {
...
-  constructor(options, input) {
+  // Forward-declarations
+  // location.js
+  +raise: (pos: number, message: string) => empty;
+  // comments.js (TODO: Better type for the parameter)
+  +addComment: (comment: Object) => void;
+  // parser/index.js
+  +hasPlugin: (name: string) => boolean;
+  // parser/util.js
+  +unexpected: (pos?: ?number, messageOrType?: string | TokenType) => empty;
```
Comment (danez) [inline src/tokenizer/index.js]: I wonder if these methods are in the wrong place currently when they are used but assign later, especially `raise` and `unexpected` and maybe other helper functions could be moved either here or into a new baseclass.
But this is out of scope for this PR.

**Comment by andy-ms** [comment]
@danez Thanks for reviewing. For type-checking the rest, would it be OK to use classes and inheritance? For example, I would like `startNode` available in `parser/index.js`, but it's currently defined with `pp.startNode = function () {`, which the type checker can't understand.

**Comment by danez** [comment]
Yes, my major concern was having the types for all parse functions duplicated, once in the function itself and once in a separate type definition. 
I thought maybe we could come up with a better structure that flow could understand and not having to duplicate type definitions, as I think the risk of forgetting to change one of them is pretty high.
But I currently don't have a good idea, besides moving everything to one file, but that is also not a good solution I think.

**Pull request merged**